libbeat/publisher/pipeline/pipeline.go

```diff
@@ -195,12 +204,21 @@ func New(
 	}
 	p.eventSema = newSema(maxEvents)
 
-	p.output = newOutputController(beat, monitors, p.observer, p.queue)
+	p.output = newOutputController(beat, monitors, p.observer, p.queue, p.getTracer)
 	p.output.Set(out)
 
 	return p, nil
 }
 
+func (p *Pipeline) getTracer() *apm.Tracer {
+	return p.tracer
+}
+
+func (p *Pipeline) SetTracer(tracer *apm.Tracer) error {
+	p.tracer = tracer
+	return nil
+}
```

Comment:
Just a hunch, but it looks like we potentially have a race condition between getTracer and SetTracer.

Reply:
certainly, this is just poc - the tracer would have to be set before publishing started like `SetACKHandler`